### PR TITLE
Support `where` with comparison operators (`>`, `>=`, `<`, and `<=`)

### PR DESCRIPTION
**Pull request by kamipo**
```ruby
posts = Post.order(:id)

posts.where("id >": 9).pluck(:id)  # => [10, 11]
posts.where("id >=": 9).pluck(:id) # => [9, 10, 11]
posts.where("id <": 3).pluck(:id)  # => [1, 2]
posts.where("id <=": 3).pluck(:id) # => [1, 2, 3]
```

From type casting and table/column name resolution's point of view,
`where("create_at >=": time)` is better alternative than `where("create_at >= ?", time)`.

```ruby
class Post < ActiveRecord::Base
  attribute :created_at, :datetime, precision: 3
end

time = Time.now.utc # => 2020-06-24 10:11:12.123456 UTC

posts = Post.order(:id)
posts.create!(created_at: time) # => #<Post id: 1, created_at: "2020-06-24 10:11:12.123000">

# SELECT `posts`.* FROM `posts` WHERE (created_at >= '2020-06-24 10:11:12.123456')
posts.where("created_at >= ?", time) # => []

# SELECT `posts`.* FROM `posts` WHERE `posts`.`created_at` >= '2020-06-24 10:11:12.123000'
posts.where("created_at >=": time) # => [#<Post id: 1, created_at: "2020-06-24 10:11:12.123000">]
```

### CODE REVIEW

File: activerecord/CHANGELOG.md
Comment (abhaynikam) [inline activerecord/CHANGELOG.md:1]: I think we should call them `comparison operators` instead of `custom`. Ref: https://www.w3resource.com/sql/comparison-operators/sql-comparison-operators.php#:~:text=A%20comparison%20(or%20relational)%20operator,two%20NULL%20expressions%20returns%20UNKNOWN).

**Comment by bogdan** [comment]
Will `!=` be supported? I know about `where.not` but it can be more readable to use `!=` when it is mixed with other conditions.

File: activerecord/CHANGELOG.md
Comment (kamipo) [reply activerecord/CHANGELOG.md:1]: 👍

**Comment by sedubois** [comment]
Could such an approach be extended to JSON operators?

```rb
posts.find_by("slugs ?": slug)
```

**Comment by kamipo** [comment]
> Will `!=` be supported? I know about `where.not` but it can be more readable to use `!=` when it is mixed with other conditions.

> Could such an approach be extended to JSON operators?
>
> ```rb
> posts.find_by("slugs ?": slug)
> ```

It is not a plan for this PR. I'd not like to expand the scope in this PR.

**Comment by simi** [comment]
What about to make this extendable (as a follow-up PR I can work on) to make possible to introduce custom operators?

**Comment by kamipo** [comment]
Seems it is fine, but not a plan for this PR.

**Comment by simi** [comment]
Providing wrong operator results into wrong SQL constructed (unknown column).

```ruby
irb(main):003:0> User.where('id ?': 1)
Traceback (most recent call last):
ActiveRecord::StatementInvalid (PG::UndefinedColumn: ERROR:  column users.id ? does not exist)
LINE 1: SELECT "users".* FROM "users" WHERE "users"."id ?" = $1 LIMI...
                                            ^
HINT:  Perhaps you meant to reference the column "users.id".
```

What about to provide custom exception "UnknownOperator" for that?

**Comment by kamipo** [comment]
Even if `where("id ?", 1)` as old syntax, it also raise StatementInvalid.

At least I'm not excited extra work in this PR to make slower things for such like unsupported usage.

**Comment by simi** [comment]
:thinking: But `where("id ?", 1)` is not the symbol, but string syntax. When passing symbol AR threats it as column name. With this nice change it threats it as column and (optional) operator. We can check if any operator is present and if that one is supported. I don't think it will slow things down. Anyway again, I can take a look once this is merged.

:mag: Looking into conflicting syntax I have found out actually `id >` is valid column name for PostgreSQL (I have not checked other DBs).

```sql
CREATE TABLE test ("id" int, "id >" int);
```

Having this table and using `Test.where('id >': 1)` would result (AFAIK) into `"id" > 1`, but actually both variants `"id >" = 1` and `"id" > 1` would be valid in here.

Anyway it is really edge case.

**Comment by kamipo** [comment]
I know "id >" is valid column name.

If we should care about that, we cannot take this idea.

At least currently all valid column names perfectly works in Active Record (e.g. non primary "id" column, "." include column name, etc).

**Comment by simi** [comment]
Cool, I was about to mention it at least for future reference. Definitely edge case Rails should not care about.

**Review by simi — APPROVED** [review]
I did some local testing, all looks good. :heart: 

I think some operators could be added (`<>`, `!=`, `=`), but as mentioned it will not be done as part of this case. I can try to open some follow-ups later.

**Comment by smridge** [comment]
Ruby ranges can also be used as an out of the box solution:
```
Post.where(id: 1..)
=>   Post Load (0.4ms)  SELECT "posts".* FROM "posts" WHERE "posts"."id" >= $1  [["id", 1]]

Post.where(id: ..9)
=>   Post Load (0.3ms)  SELECT "posts".* FROM "posts" WHERE "posts"."id" <= $1  [["id", 9]]

Post.where(id: ...9)
=>   Post Load (0.3ms)  SELECT "posts".* FROM "posts" WHERE "posts"."id" < $1  [["id", 9]]
```

**Comment by tenderlove** [comment]
This is an interesting change, but since this changes public API I think we need to have a discussion with the rest of the core team before merging.  Bugfixes and performance improvements are fine for merging without discussion, but this type of change needs a bit more scrutiny by the rest of the core team.  Can we please revert this and discuss more before merging?

**Comment by kamipo** [comment]
Sure, I will be going to revert this.

I'd recommend to people use the hash syntax, the hash syntax also have other useful effects (making boundable queries, unscopeable queries, hash-like relation merging friendly, automatic other table references detection, let me explain in a later PR).

Hopefully the hash syntax would be approved from the core team...

**Pull request merged**
